### ISSUE TITLE
Shapes demo fails with RTI 6.1 on Ubuntu 18.04

### ISSUE THREAD

**Issue by maspe36** [jhelovuo/RustDDS#147]
Like the title says, I'm unable to hear anything back from the RTI shapes demo. I'm publishing a blue square with the default QoS settings.

rustdds version 0.6.2
Command to run rustdds `shapes_demo` example
`RustDDS/examples/shapes_demo$ cargo run --example=shapes_demo -- -S -t Square`

`rtishapesdemo` output (Version 6.1)
```
$ ./rtishapesdemo 
Gtk-Message: 12:21:19.629: Failed to load module "canberra-gtk-module"
DL Info: : Domain participant created on domain 0
DL Info: : A new generic shape has been created
DL Info: : DataWriterQoS set to use Default Profile
DL Info: : DataReaderQoS set to use Default Profile
DL Info: : <SHAPE CREATED> Created Writer shape Square.
DL Info: : <-PUBLISHER-> A new square publisher has been successfully created
DL Debug: : 	Backtrace:
DL Debug: : 	#4	DDS_TypeLibraryElementSeq_ensure_length ??:? [0x103E9B6]
DL Debug: : 	#5	DDS_DynamicDataStream_print_container ??:? [0xE9671B]
DL Debug: : 	#6	DDS_DynamicDataStream_equal_subtype ??:? [0xEA2BE7]
DL Debug: : 	#7	DDS_DynamicDataStream_get_size_member ??:? [0xEA3723]
DL Debug: : 	#8	DDS_DomainParticipantSeq_get_reference ??:? [0xE582B2]
DL Debug: : 	#9	DDSFactoryPluginSupport::createParticipant(int, DDS_DomainParticipantQos const*, void*, unsigned int) ??:? [0xD3A111]
DL Debug: : 	#10	RTI_DL_ADMIN_RTILoggerVerbositySeq_loan_contiguous ??:? [0xD0FFC4]
DL Debug: : 	#11	RTI_DL_ADMIN_RTILoggerVerbositySeq_loan_contiguous ??:? [0xD0FFC4]
DL Debug: : 	#12	RTI_DL_ADMIN_RTILoggerCategorySeq_get ??:? [0xD0E9B3]
DL Debug: : 	#13	start_thread /build/glibc-S9d2JN/glibc-2.27/nptl/pthread_create.c:463 [0x76DB]
DL Debug: : 	#14	clone /build/glibc-S9d2JN/glibc-2.27/misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:97 [0x12171F]
DL Fatal: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000200C2{E=DW,T=PRESInterParticipantTopic,C=PRESInterParticipantParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000200C7{C=PRESInterParticipantParameter}] Mx00:/rti/jenkins/workspace/connextdds/6.1.0.0/x64Linux2.6gcc4.4.5/src/commend.1.0/srcC/srw/SrWriterService.c:16642:RTI0x2000026:!precondition
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000200C2{E=DW,T=PRESInterParticipantTopic,C=PRESInterParticipantParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000200C7{C=PRESInterParticipantParameter}] PRESPsService_assertRemoteReaderInLocalSessionWriter:!srw->assertRemoteReader
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000200C2{E=DW,T=PRESInterParticipantTopic,C=PRESInterParticipantParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000200C7{C=PRESInterParticipantParameter}] PRESPsService_assertCompatibleLocalWriterRemoteReaderMatchLinkState:!assert remote reader in local session writer
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000200C2{E=DW,T=PRESInterParticipantTopic,C=PRESInterParticipantParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000200C7{C=PRESInterParticipantParameter}] PRESPsService_linkLocalWriterToRemoteReader:!assert compatible local writer remote reader link state
DL Error: : PRESPsService_linkRemoteReaderToLocalWriters:[Local Endpoint: 1018b4f 72998b9a cf2472af 200c2] [Remote Endpoint: 2270b152 abf74f76 8e8117d0 200c7] failed to link local writer to remote reader
DL Error: : PRESPsService_onLinkToLocalEndpointEvent:!create linkRemoteReaderToWriters
DL Fatal: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000004C2{E=DW,T=DISCSubscription,C=DISCSubscriptionParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000004C7{C=DISCSubscriptionParameter}] Mx00:/rti/jenkins/workspace/connextdds/6.1.0.0/x64Linux2.6gcc4.4.5/src/commend.1.0/srcC/srw/SrWriterService.c:16642:RTI0x2000026:!precondition
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000004C2{E=DW,T=DISCSubscription,C=DISCSubscriptionParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000004C7{C=DISCSubscriptionParameter}] PRESPsService_assertRemoteReaderInLocalSessionWriter:!srw->assertRemoteReader
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000004C2{E=DW,T=DISCSubscription,C=DISCSubscriptionParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000004C7{C=DISCSubscriptionParameter}] PRESPsService_assertCompatibleLocalWriterRemoteReaderMatchLinkState:!assert remote reader in local session writer
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000004C2{E=DW,T=DISCSubscription,C=DISCSubscriptionParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000004C7{C=DISCSubscriptionParameter}] PRESPsService_linkLocalWriterToRemoteReader:!assert compatible local writer remote reader link state
DL Error: : PRESPsService_linkRemoteReaderToLocalWriters:[Local Endpoint: 1018b4f 72998b9a cf2472af 4c2] [Remote Endpoint: 2270b152 abf74f76 8e8117d0 4c7] failed to link local writer to remote reader
DL Error: : PRESPsService_onLinkToLocalEndpointEvent:!create linkRemoteReaderToWriters
DL Fatal: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000003C2{E=DW,T=DISCPublication,C=DISCPublicationParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000003C7{C=DISCPublicationParameter}] Mx00:/rti/jenkins/workspace/connextdds/6.1.0.0/x64Linux2.6gcc4.4.5/src/commend.1.0/srcC/srw/SrWriterService.c:16642:RTI0x2000026:!precondition
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000003C2{E=DW,T=DISCPublication,C=DISCPublicationParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000003C7{C=DISCPublicationParameter}] PRESPsService_assertRemoteReaderInLocalSessionWriter:!srw->assertRemoteReader
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000003C2{E=DW,T=DISCPublication,C=DISCPublicationParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000003C7{C=DISCPublicationParameter}] PRESPsService_assertCompatibleLocalWriterRemoteReaderMatchLinkState:!assert remote reader in local session writer
DL Error: : [0x01018B4F,0x72998B9A,0xCF2472AF:0x000003C2{E=DW,T=DISCPublication,C=DISCPublicationParameter,D=0}|LINK 0x2270B152,0xABF74F76,0x8E8117D0:0x000003C7{C=DISCPublicationParameter}] PRESPsService_linkLocalWriterToRemoteReader:!assert compatible local writer remote reader link state
DL Error: : PRESPsService_linkRemoteReaderToLocalWriters:[Local Endpoint: 1018b4f 72998b9a cf2472af 3c2] [Remote Endpoint: 2270b152 abf74f76 8e8117d0 3c7] failed to link local writer to remote reader
DL Error: : PRESPsService_onLinkToLocalEndpointEvent:!create linkRemoteReaderToWriters

```

RTI to RTI connection is fine, and the errors only appear when I'm starting up the rustdds example.

I am able to subscribe to the eProsima's ShapeDemo (v2.4.0) Square topic created with the same QoS settings. 

EDIT: Seems like rustdds 0.6.1 can actually communicate with the RTI 6.1 shapes demo. Order of DomainParticipant creation still matters as described [here](https://github.com/jhelovuo/RustDDS/issues/142).

**Comment by jhelovuo** [comment]
Since the order of DomainParticipant creation matters, maybe this is related somehow to competing on the same resources, such as UDP port numbers?

How did you get the error log from RTI Connext Shapes Demo? Even though error log is better than nothing, and things seem to go wrong on three lines beginning with "DL Fatal", the error message is a bit cryptic and it is unclear what is the matter here. It says some precondition is not satisfied, but which one?

Did you try this with both RTI and RustDDS running on the same host? I suppose running on different hosts in the same LAN works ok?

Additionally, I just checked against RTI ShapesDemo 5.3 on Linux, both running on the same host, and RustDDS talks just fine to Connext 5.3, regardless of the initialization order.

**Comment by jhelovuo** [comment]
> Additionally, I just checked against RTI ShapesDemo 5.3 on Linux, both running on the same host, and RustDDS talks just fine to Connext 5.3, regardless of the initialization order.

On closer inspection, the RTI ShapesDemo works against RustDDS 0.6.1, but fails against 0.6.2, so this is a regression in 0.6.2. Exact cause is still under investigation.

**Comment by maspe36** [comment]
> How did you get the error log from RTI Connext Shapes Demo?
I just ran the shapes demo directly from the CLI and that was in the stderr/stdout

> Since the order of DomainParticipant creation matters, maybe this is related somehow to competing on the same resources, such as UDP port numbers?

Yeah I'm also curious about this, I'm not super familiar with different DDS providers - but it does seem like RTI heavily leverages shared memory. 

> Did you try this with both RTI and RustDDS running on the same host? I suppose running on different hosts in the same LAN works ok?

Yeah same host, don't have another host readily available to test with so I can't confirm it would work on another host in the same LAN.

> On closer inspection, the RTI ShapesDemo works against RustDDS 0.6.1, but fails against 0.6.2, so this is a regression in 0.6.2. Exact cause is still under investigation.

Gotcha, I noted this in my edit but I should've added it as an additional comment. Let me know if there is anything I can do to help root cause or contribute!

**Comment by jhelovuo** [comment]
RustDDS was accidentally sending out normal IP addresses as multicast group addresses in discovery conversation, which Connext (correctly) flagged as an error. This should now be fixed in version 0.6.3.

@maspe36 , does it work for you?

**Comment by maspe36** [comment]
Yes confirmed 0.6.3 works with the shapes demo as a subscriber/publisher with out depending on the creation order of the DomainParticipant!
